packages/roosterjs-editor-core/lib/coreApi/getPendingFormatState.ts

```diff
@@ -0,0 +1,71 @@
+import { DocumentCommand, NodePosition, PendableFormatState } from 'roosterjs-editor-types';
+import { PendableFormatCommandMap, PendableFormatNames, Position } from 'roosterjs-editor-dom';
+
+/**
+ *
+ * @param range The range of the cursor.
+ * @param cachedPendableFormatState The format state cached by PendingFormatStatePlugin.
+ * @param cachedPosition The position cached by PendingFormatStatePlugin.
+ * @returns The cached format state if it exists. If the cached postion do not exist, search for    pendable elements in the DOM tree and return the pendable format state.
+ */
+
+export function getPendingFormatState(
+    range: Range,
+    cachedPendableFormatState: PendableFormatState,
+    cachedPosition: NodePosition
+): PendableFormatState {
+    const currentPosition = range && Position.getStart(range).normalize();
+    const isSamePosition = range && currentPosition.equalTo(cachedPosition);
+    if (range && cachedPendableFormatState && range.collapsed && isSamePosition) {
+        return cachedPendableFormatState;
+    } else {
+        return getPendableFormatState();
+    }
+}
+
+function isAPendableFormat(element: HTMLElement, format: string): boolean {
+    switch (format) {
```

Comment:
For these checking function, how about make them as a function map:
```
const StyleCheckers: Record<PendableFormatNames, (tagName: string, style: CSSStyleDeclaration) => boolean> {
 isBold: (tag, style) => tag == 'B' || tag == 'STRONG' || style.fontWeight >= 700,
 ...
};
```

Then you just need a very simple loop to check all the styles:
```
  while (contains(core.contentDiv, node)) {  // We just need to check the node under editor, no need to loop to root node of the page
     const tag = getTagOfNode(node);
     const style = node.nodeType == NodeType.Element && (<Element>node).style;

     if (tag & style) { // Only need to check for Element
        Object.keys(StyleCheckers).forEach((style: PendableFormatNames) => {
           formatState[style] = formatState[style] || StyleCheckers[style](tag, style);
        });
    }
    node = node.parentNode; // Let's use parentNode instead of parentElement since parentElement may have some issue in IE
  }
```